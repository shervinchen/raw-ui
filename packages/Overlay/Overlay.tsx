import React, { FC, MouseEvent } from 'react';
import { OverlayProps } from './Overlay.types';
import { useTheme } from '../Theme';
import { useTransition } from '../utils/hooks';

const Overlay: FC<OverlayProps> = ({ visible, onClick, ...restProps }) => {
  const theme = useTheme();
  const { stage, shouldMount } = useTransition(visible, 50, 350);

  const clickHandler = (event: MouseEvent<HTMLDivElement>) => {
    onClick?.(event);
  };

  return (
    shouldMount && (
      <div
        data-testid="overlay"
        className="raw-overlay"
        onClick={clickHandler}
        style={{
          opacity: stage === 'enter' ? 0.25 : 0,
        }}
        {...restProps}
      >
        <style jsx>{`
          .raw-overlay {
            position: fixed;
            top: 0;
            left: 0;
            right: 0;
            bottom: 0;
            width: 100%;
            height: 100%;
            z-index: 1000;
            background-color: ${theme.palette.foreground};
            transition: opacity 0.35s cubic-bezier(0.4, 0, 0.2, 1);
          }
        `}</style>
      </div>
    )
  );
};

export default Overlay;
