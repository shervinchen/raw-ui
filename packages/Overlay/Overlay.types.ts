import { ComponentPropsWithoutRef, MouseEvent } from 'react';
import { Merge } from '../utils';

export interface BaseOverlayProps {
  visible?: boolean;
  onClick?: (event: MouseEvent<HTMLDivElement>) => void;
}

export type OverlayProps = Merge<
  ComponentPropsWithoutRef<'div'>,
  BaseOverlayProps
>;
