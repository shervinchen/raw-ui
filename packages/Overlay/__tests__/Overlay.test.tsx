import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Overlay from '..';

describe('Overlay', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(<Overlay visible />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should trigger click event', async () => {
    const mockClickFn = jest.fn();
    const { getByTestId } = render(<Overlay visible onClick={mockClickFn} />);
    await userEvent.click(getByTestId('overlay'));
    expect(mockClickFn).toHaveBeenCalledTimes(1);
  });
});
