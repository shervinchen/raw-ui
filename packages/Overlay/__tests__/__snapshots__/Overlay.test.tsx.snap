// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-251717942 raw-overlay"
    data-testid="overlay"
    style="opacity: 0;"
  />
</DocumentFragment>
`;
